src/core/or/circuitbuild.c

```diff
@@ -2418,17 +2414,29 @@ count_acceptable_nodes, (smartlist_t *nodes))
     if (! node->is_valid)
 //      log_debug(LD_CIRC,"Nope, the directory says %d is not valid.",i);
       continue;
-    if (! node_has_any_descriptor(node))
+    /* If we can use the node, check if we can connect directly. */
+    if (node_has_preferred_descriptor(node, 1))
```

Comment:
Here you treat "indirect" and "direct" as having a logical relation to one another, as if it's okay to count every "direct" node as also being "indirect" later on (when you do direct+indirect).  That isn't exactly the case, though: for the nodes in the "direct" count, we might not be willing to use them indirectly if we don't have the right descriptor for them.

Probably direct and indirect should be computed separately; maybe "direct" should be an argument to this function?

Reply:
It's also worth noting that bridges are used as entry nodes (direct) but not middle or end nodes (indirect).
The one exception is when a public relay is set as a bridge.

Reply:
To make a path of length N, tor needs at least 1 direct entry node (guard, or bridge, or any node if not using entry guards) and at least N-1 indirect non-entry nodes.

We need to replace the simple test in new_route_len() with separate tests for direct and indirect, and update the log message as well.